### PR TITLE
[#2199] Pressing delete while a trackevent is selected now closes the open editor and then reopens the plugin list

### PR DESCRIPTION
**Pull request by mjschranz**
https://webmademovies.lighthouseapp.com/projects/65733/tickets/2199-when-a-track-event-is-deleted-its-editor-should-be-closed

### CODE REVIEW

File: src/editor/module.js [hunk 1]
```diff
@@ -72,6 +72,19 @@ define( [ "core/eventmanager", "core/trackevent", "./editor",
...
 
     /**
+     * Member: closeEditor
+     *
+     * Closes the currently opened editor and opens the default one in it's place
+     *
+     */
+    _this.closeEditor = function() {
+      if( _currentEditor ) {
+        _currentEditor.close();
```
Comment (secretrobotron) [inline src/editor/module.js]: To avoid `close` being called twice, set `_currentEditor` to null here.

File: src/ui/ui.js [hunk 2]
```diff
@@ -250,9 +250,13 @@ define( [ "core/eventmanager", "./toggler",
...
         if( butter.selectedEvents.length ) {
+          var trackEvent,
+              i;
           e.preventDefault();
-          for( var i = 0; i < butter.selectedEvents.length; i++ ) {
-            butter.selectedEvents[ i ].track.removeTrackEvent( butter.selectedEvents[ i ] );
+          for( i = 0; i < butter.selectedEvents.length; i++ ) {
+            trackEvent = butter.selectedEvents[ i ];
+            trackEvent.track.removeTrackEvent( trackEvent );
+            butter.editor.closeEditor();
```
Comment (secretrobotron) [inline src/ui/ui.js]: Don't want this to happen for every event in the loop.

File: src/ui/ui.js [hunk 3]
```diff
@@ -250,9 +250,13 @@ define( [ "core/eventmanager", "./toggler",
       }, // esc key
       8: function( e ) { // del key
         if( butter.selectedEvents.length ) {
+          var trackEvent,
+              i;
           e.preventDefault();
-          for( var i = 0; i < butter.selectedEvents.length; i++ ) {
-            butter.selectedEvents[ i ].track.removeTrackEvent( butter.selectedEvents[ i ] );
+          for( i = 0; i < butter.selectedEvents.length; i++ ) {
+            trackEvent = butter.selectedEvents[ i ];
```
Comment (secretrobotron) [inline src/ui/ui.js]: Why the extra variable `trackevent`?

Comment (mjschranz) [inline src/ui/ui.js]: A relic from something else I tried originally. Being removed.

File: src/editor/base-editor.js [hunk 4]
```diff
@@ -68,7 +68,9 @@ define( [ "core/eventmanager", "util/scrollbars", "ui/widget/tooltip" ],
      */
     extendObject.close = function() {
       // Remove the editor's root element from the element to which it was attached
-      extendObject.rootElement.parentNode.removeChild( extendObject.rootElement );
+      if ( extendObject.rootElement.parentNode ) {
```
Comment (secretrobotron) [inline src/editor/base-editor.js]: With the change you made here https://github.com/mozilla/butter/pull/938/files#L1R81, this check might be unnecessary. I'd rather have this function fail as an obvious sign that we call `close` way too much than it half-fail silently.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @dseif received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/fc3251cb4701530/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/fc3251cb4701530/output.txt

Total script time: 1.46 mins

Server is running on http://pr938.butyr.org/templates/basic/
Run the Test Suite: http://pr938.butyr.org/test/

File: src/editor/module.js [hunk 5]
```diff
@@ -184,7 +196,7 @@ define( [ "core/eventmanager", "core/trackevent", "./editor",
           butter.loader.load( editorsToLoad, function() {
             Editor.loadUrlSpecifiedLayouts( onModuleReady, butter.config.value( "baseDir" ) );
           }, function( e ) {
-            _logger.log( "Couldn't load editor " + e.srcElement.src );
```
Comment (cadecairos) [inline src/editor/module.js:187]: srcElement eh.... :P

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @cadecairos received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/29a78af227759a1/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/29a78af227759a1/output.txt

Total script time: 0.77 mins

Server is running on http://pr938.butyr.org/templates/basic/
Run the Test Suite: http://pr938.butyr.org/test/

File: src/editor/base-editor.js [hunk 6]
```diff
@@ -74,7 +74,7 @@ define( [ "core/eventmanager", "util/scrollbars", "ui/widget/tooltip" ],
       if ( events.close ) {
         events.close.apply( extendObject, arguments );
       }
-      
+
```
Comment (cadecairos) [inline src/editor/base-editor.js:77]: revert this line change

File: src/editor/trackevent-editor.js [hunk 7]
```diff
@@ -33,7 +33,8 @@ define([ "util/lang", "util/keys", "util/time", "./base-editor",
    */
   return function( extendObject, butter, rootElement, events ) {
     // Wedge a check for scrollbars into the open event if it exists
-    var oldOpenEvent = events.open;
+    var oldOpenEvent = events.open,
+        _trackEvent;
```
Comment (cadecairos) [inline src/editor/trackevent-editor.js:37]: use no underscores or underscores, but not both

File: src/editor/trackevent-editor.js [hunk 8]
```diff
@@ -585,6 +588,24 @@ define([ "util/lang", "util/keys", "util/time", "./base-editor",
...
     };
 
+    extendObject.getTrackEvent = function() {
+      return _trackEvent;
+    };
+
+    butter.listen( "trackeventremoved", function closeEditor( e ) {
+      butter.unlisten( "trackeventremoved", closeEditor );
+
+      var currentTrackEvent;
```
Comment (cadecairos) [inline src/editor/trackevent-editor.js]: hahahaha

File: src/editor/base-editor.js (hunk 6 above)
Comment (mjschranz) [inline src/editor/base-editor.js:77]: I don't know why that appears. That's exactly how the file is on master.

**Pull request merged**